src/cpp/session/SessionSourceDatabase.cpp

```diff
@@ -716,6 +732,24 @@ void onShutdown(bool)
       LOG_ERROR(error);
 }
 
+void onDocUpdated(boost::shared_ptr<SourceDocument> pDoc)
+{
+   s_idToPath[pDoc->id()] = pDoc->path();
+}
+
+void onDocRemoved(const std::string& id)
```

Comment:
There is also an onRemoveAll that you need to handle (this can of course all be done in the centralized implementation)